java/src/main/java/io/spine/validation/java/JavaValidationRenderer.java

```diff
@@ -142,17 +190,34 @@ private CodeBlock generateValidationCode(MessageValidation validation,
             GenerationContext context = new GenerationContext(
                     rule, result, file, typeSystem, typeName, VIOLATIONS, this
             );
-            JavaCodeGenerator generator = JavaCodeGeneration.generatorFor(context);
+            CodeGenerator generator = JavaCodeGeneration.generatorFor(context);
             CodeBlock block = generator.code();
             code.add(block);
         }
         return code.build();
     }
 
-    private static CodeBlock throwValidationException() {
+    private static CodeBlock extraInsertionPoint(TypeName name) {
+        InsertionPoint insertionPoint = new ExtraValidation(name);
+        Language java = CommonLanguages.java();
+        String line = java.comment(getCodeLine(insertionPoint)) + lineSeparator();
+        return CodeBlock.of(line);
+    }
+
+    private static CodeBlock generateValidationError() {
         CodeBlock.Builder code = CodeBlock.builder();
         code.beginControlFlow("if (!$N.isEmpty())", VIOLATIONS);
-        code.addStatement("throw new $T($N)", ValidationException.class, VIOLATIONS);
+        MethodCall errorBuilder = new ClassName(ValidationError.class)
+                .newBuilder()
+                .chainAddAll("constraint_violation", new Literal(VIOLATIONS))
+                .chainBuild();
+        MethodCall newOptional = new ClassName(Optional.class)
+                .call("of", newArrayList(errorBuilder), newArrayList());
```

Comment:
If I remember correctly, `newArrayList()` is deprecated for Java starting 7th. Please review.